packages/cdktf/lib/terraform-functions.ts

```diff
@@ -0,0 +1,1221 @@
+import { Tokenization } from "./tokens/token";
+import { call } from "./tfExpression";
+import { IResolvable } from "./tokens/resolvable";
+import { TokenMap } from "./tokens/private/token-map";
+
+// We use branding here to ensure we internally only handle validated values
+// this allows us to catch usage errors before terraform does in some cases
+type TFValue = any & { __type: "tfvalue" };
+type TFValueValidator = (value: any) => TFValue;
+
+type ExecutableTfFunction = (...args: any[]) => IResolvable;
+
+function anyValue(value: any): any {
+  return value;
+}
+
+function mapValue(value: any): any {
+  return value;
+}
+
+function stringValue(value: any): any {
+  if (typeof value !== "string" && !Tokenization.isResolvable(value)) {
+    throw new Error(`${value} is not a valid number nor a token`);
+  }
+  return value;
+}
+
+function numericValue(value: any): any {
+  if (typeof value !== "number" && !Tokenization.isResolvable(value)) {
+    throw new Error(`${value} is not a valid number nor a token`);
+  }
+  return value;
+}
+
+function listOf(type: TFValueValidator): TFValueValidator {
+  return (value: any) => {
+    if (Tokenization.isResolvable(value)) {
+      return value;
+    }
+
+    if (!Array.isArray(value)) {
+      //   throw new Error(`${value} is not a valid list`);
+      return value;
+    }
+
+    value.forEach((item, i) => {
+      if (Tokenization.isResolvable(item)) {
+        return;
+      }
+
+      try {
+        type(item);
+      } catch (error) {
+        throw new Error(
+          `Element in list ${value} at position ${i} is not of the right type: ${error}`
+        );
+      }
+    });
+
+    return value;
+  };
+}
+
+function asString(value: IResolvable) {
+  return TokenMap.instance().registerString(value);
+}
+
+function asNumber(value: IResolvable) {
+  return TokenMap.instance().registerNumber(value);
+}
+
+function asList(value: IResolvable) {
+  return TokenMap.instance().registerList(value);
+}
+
+function terraformFunction(
+  name: string,
+  argValidators: TFValueValidator | TFValueValidator[]
+): ExecutableTfFunction {
+  return function (...args: any[]) {
+    if (Array.isArray(argValidators)) {
+      if (args.length !== argValidators.length) {
+        throw new Error(
+          `${name} takes ${argValidators.length} arguments, but ${args.length} were provided`
+        );
+      }
+      args.forEach((arg, i) => argValidators[i](arg));
+      return call(name, args);
+    } else {
+      return call(name, argValidators(args));
+    }
+  };
+}
+
+export class Fn {
```

Comment:
Might be nice to split this into multiple files.

Reply:
I tried, but JSII exploded then 🙈  Had them all namespaced in their group but JSII ran into problems